Feat/media3 newarch

jspizziri opened the pull request:
- Migrates the library to use media3 on Android.
- Migrates the library to be compatible with the newArch.
- Removes support for the legacy architecture.
- Removes deprecated functionality.


#### TODO:

- [ ] Add web example
- [x] Fix issues with Remote playback events not firing.
- [x] Test web
- [x] Upgrade example to `react-native@0.80.0`
- [x] Fix iOS CI pipeline

behenate committed:
chore: improve bridgeless interop layer compatibility

robik committed:
fix: broken events in bridgeless mode (React Native New Architecture)

puckey committed:
chore: update node to v22

Node 22 has entered LTS as of October 2024

puckey committed:
example: update to react-native 0.76.1

example: fix

puckey committed:
android: inline KotlinAudio

lovegaoshi committed:
fix: catch currentTrack null

puckey committed:
hooks: fix circular dependency

puckey committed:
chore: remove deprecated functionalities

puckey committed:
feat: initial media3 functionality

Brings in media3 & android-auto functionalities from https://github.com/lovegaoshi/react-native-track-player, leaving out widget & audio fading functionalities

puckey committed:
android: various

- show jump backward & forward buttons in notification
- remove compactCapabilities
- rename BaseAudioPlayer#player to BaseAudioPlayer#forwardingPlayer

puckey committed:
example: update

lovegaoshi committed:
feat: onStartCommandIntentValid

lovegaoshi committed:
feat: onStartCommandIntentValid

puckey committed:
android: remove forwarding player

puckey committed:
example: add android:adb-reverse script

puckey committed:
example: update react-native to 0.76.2

puckey committed:
android: default mediaId to audioItem.audioUrl

puckey committed:
android: rename timber tags to RNTP

puckey committed:
remove deprecated clearNowPlayingMetadata method

See https://github.com/doublesymmetry/react-native-track-player/issues/1653

puckey committed:
android: rename variable mediaSource to mediaItem in replaceItem method

puckey committed:
android: inline FocusManager & bring back ForwardingPlayer

puckey committed:
android: simplify getActiveTrack

puckey committed:
android: allow currentTrack to be null

puckey committed:
android: forward custom commands to forwarding player

puckey committed:
android: fix updateMetadataForTrack

puckey committed:
android: improve logging

puckey committed:
android: interceptPlayerActionsTriggeredExternally

puckey committed:
android: rename inner class

puckey committed:
android: bring back rating support

puckey committed:
android: remove deprecated isForegroundService

puckey committed:
android: clean up

puckey committed:
android: remove custom actions for now

puckey committed:
android: replace Log with Timber for logging

puckey committed:
android: remove embedded bitmap functionality for now

puckey committed:
android: refactor audioItem2MediaItem to use coroutines for artwork loading

puckey committed:
android: turn AudioItem into a class

android: update TrackAudioItem to initialize AudioItem properties

puckey committed:
android: remove embedded bitmap remnants

puckey committed:
android: fix rating class

puckey committed:
remove custom buttons for now

puckey committed:
android: rename cache directory from APM to RNTP

puckey committed:
android: further refactor audio items

Ken-Michalak committed:
Merge branch 'feature/media3' of github.com:puckey/react-native-track-player into feature/media3

Ken-Michalak committed:
style: fix format and ci validate errors

jspizziri committed:
feat(android, ios): migrate the library to the new architecture

jspizziri commented:
@tsipho description updated.

jspizziri committed:
Merge branch 'main' into feat/media3-newarch

jspizziri committed:
resolve conflicts

jspizziri committed:
update lock

jspizziri committed:
chore: fix build errors

jspizziri committed:
fix(ios): fix a build error by removing deprecated event listener

jspizziri committed:
chore(example): upgrade example to react-native@0.80.0

jspizziri committed:
fix(android): resolve build errors in react-native@0.80.0

jspizziri committed:
chore(example): replace safe-area with react-native-save-area-context

jspizziri committed:
docs(getting-started): remove use of deprecated and removed method

jspizziri commented:
@janwiebe-jump this PR removes support for the oldArch, we will only be supporting newArch moving forward.

jspizziri commented:
@thedv91 ive not seen this. Please repro in the example app.

sonnv1912 commented:
> It works fine on iOS but on Android it throws `java.lang.NullPointerException: java.lang.NullPointerException` when calling `TrackPlayer.add`. Does anyone have this error like me?

same

jspizziri commented:
@thedv91 @sonnv1912 I need a repro in the _example_ app.

sonnv1912 commented:
> https://github.com/thedv91/track-player-sample

@thedv91 has already created example repo here.

[https://github.com/thedv91/track-player-sample](https://github.com/thedv91/track-player-sample/blob/main/app/(tabs)/index.tsx)

jspizziri commented:
@sonnv1912 , I'm not asking for a repro in "an example",but rather "in the example app": https://github.com/doublesymmetry/react-native-track-player/tree/feat/media3-newarch/example

I want a reproduction in the example app in this repo, there are too many external variables outside of the scope of this repo in someone else's hand spun repro app.

radko93 commented:
Tested on RN 0.79.5 (expo modules installed but it's not an expo app. All good.

subvertallchris commented:
Working great in latest Expo with iOS 18.5

ega65 commented:
Guys, just to say thanks to your effort to port this nice piece of software to Expo53 newarch..  ..i'm stuck in expo51 with it, on a radio app developed with no charge, have a 1000 downloads,...  ...if i had the knowledge to contribute, i would do it without doubt...

hennessyevan commented:
I tested it in my app with expo53 and the new arch enabled. While it worked, not all of the events fired. Specifically playbackState, not in the app or in the service for remote play or pause.

Playback Progress seemed to come through fine though.

jspizziri commented:
@hennessyevan @thedv91 

If you'd like us to look into your expo concerns please open a PR into this branch that adds an "example-expo" folder which:

1. Properly installs RNTP from the parent dir.
2. Reproduces your issues.

vmsilva commented:
Same issue here i've been clone this branch  install dependencies and build but, when I try link with yarn i've got an error.
My project is on Expo SDK 53, newArch, RN 0.79.5, bare workflow
Any solution?

doughsay commented:
I have tested this branch in my app on expo 53/RN 0.79.5 with new arch enabled, and at first glance everything seems to be working great.  Background service and events are firing just fine. I haven't 100% thoroughly tested it, but I would be happy if you published at least an RC to npm, that would make using it in builds a lot easier.

I'm not sure why others are having issues with expo...  here's all I did:

* clone RNTP
* switch to this branch
* yarn
* yarn build
* yarn pack
* then in my app: expo install <path-to-tgz>

Bogdastotel commented:
when can we expect this pr to be approved?

doughsay commented:
Ok, yeah I can also reproduce the same crash when calling `TrackPlayer.add` while having the `maxCacheSize` setting enabled.  But it only happens for streaming URLs for me, if I load a locally downloaded file using `expo-file-system` it doesn't crash.

deveshp commented:
Checked out a few different streaming radios and some seem to not be working on android.

Example of streams that work : https://icecast.walmradio.com:8443/classic, https://mangoradio.stream.laut.fm/mangoradio

Example of streams that don't : https://airhlspush.pc.cdn.bitgravity.com/httppush/hlspbaudio005/hlspbaudio005_Auto.m3u8, http://stream.live.vc.bbcmedia.co.uk/bbc_world_service

deveshp commented:
> Checked out a few different streaming radios and some seem to not be working on android.
> 
> Example of streams that work : https://icecast.walmradio.com:8443/classic, https://mangoradio.stream.laut.fm/mangoradio
> 
> Example of streams that don't : https://airhlspush.pc.cdn.bitgravity.com/httppush/hlspbaudio005/hlspbaudio005_Auto.m3u8, http://stream.live.vc.bbcmedia.co.uk/bbc_world_service

Alright, used these steps, and now the streams load except for the m3u stream.
```javascript
  await TrackPlayer.reset();
  await TrackPlayer.load(track);
  await TrackPlayer.play();
```

ega65 commented:
Hello Guys, I would like to help and test it on expo custom dev, but the only way i know to install a library is with npm install or npx expo install.. ..any clue how to replace my RNTP 4 by this what seems to be a nice update ? ;)

vmsilva commented:
> Hello Guys, I would like to help and test it on expo custom dev, but the only way i know to install a library is with npm install or npx expo install.. ..any clue how to replace my RNTP 4 by this what seems to be a nice update ? ;)

I've created an npm lib to use till next update, here url https://www.npmjs.com/package/@vmsilva/react-native-track-player

aalekz commented:
Using this in my app running Expo 53 and works great with the new architecture, great work!

I have one question about this commit though: https://github.com/doublesymmetry/react-native-track-player/commit/b11736b7f249698730845339c086b7dc722ca1ee 

It removes both the code emitting `MetadataTimedReceived` events  as well as the deprecated `EventType.PlaybackMetadataReceived`. Because of this I'm having issues getting the in stream metadata, by adding back the timed metadata event listener on init and a simpler version of `handleAudioPlayerTimedMetadataReceived` removing the usage of the deprecated `EventType.PlaybackMetadataReceived` fixes the issue.  See example below:

```
// Trackplayer init()
player.event.receiveTimedMetadata.addListener(self, handleAudioPlayerTimedMetadataReceived)

func handleAudioPlayerTimedMetadataReceived(metadata: [AVTimedMetadataGroup]) {
        let metadataItems = MetadataAdapter.convertToGroupedMetadata(metadataGroups: metadata);
        emit(event: EventType.MetadataTimedReceived, body: ["metadata": metadataItems])
}
```

I believe this might have been an unintended change, but not sure :)

dodgex commented:
Is it intended, that `compactCapabilities` has been removed from `UpdateOptions`? I wonder especially, as the [`docs/versioned_docs/version-5.0/api/objects/update-options.md`](https://github.com/doublesymmetry/react-native-track-player/pull/2478/files#diff-6f336db514fc55e11d065a50080c2f8a335bc711115911d92bf9e669157bcffa) and [`docs/versioned_docs/version-5.0/basics/getting-started.md`](https://github.com/doublesymmetry/react-native-track-player/pull/2478/files#diff-afeae297475a9ca95c0480236585c15a31cb7bf031ad9bd5c95712ee62b69a98) still mention that field as if it still is supported.

jspizziri committed:
chore(android): remove references to android options that are no longer supported

jspizziri commented:
@ngambmicheal @doughsay , I've confirmed the issue with `maxCacheSize`. I'll try to get to the bottom of it.

lrobak0 commented:
Hi, I noticed that the following event listeners are not working as expected:  

TrackPlayer.addEventListener(Event.RemotePause, () => { ... });
TrackPlayer.addEventListener(Event.RemotePlay, () => { ... });

They don't seem to be triggered when interacting with the system media controls. (play/pause buttons)

jspizziri committed:
fix(android): fix an issue with maxCacheSize option on android causing crashes

jspizziri committed:
chore(ts): remove unused event type and fix deprecation warning

jspizziri commented:
@ngambmicheal @doughsay I believe that https://github.com/doublesymmetry/react-native-track-player/pull/2478/commits/349eff03b930bc85da2a21c7e40bd16764989ef0 should fix your issue. Please test and LMK.

jspizziri commented:
@deveshp , I think there's likely some errors with your streams. All the streams we have configured in the test app are working like they have prior to this change. You might try debugging the sources themselves for issues, and also testing them in the `main` example app to confirm the issue is or is not present outside of this branch.

jspizziri committed:
feat(web): migrate web to turbomodule architecture

lrobak0 commented:
> > Hi, I noticed that the following event listeners are not working as expected:
> > TrackPlayer.addEventListener(Event.RemotePause, () => { ... });
> > TrackPlayer.addEventListener(Event.RemotePlay, () => { ... });
> > They don't seem to be triggered when interacting with the system media controls. (play/pause buttons)
> 
> See my message above. I think you need to use PlayPause

...also doesn't work. I'm using react native CLI and tested it on Android 15

jspizziri commented:
@doughsay ,

> Unfortunately no, it seems to still crash for me when using caching. I get [java.lang.NullPointerException: java.lang.NullPointerException] and having a hard time getting any more detail than that. I wasn't actually using the maxCacheSize option, so this doesn't affect me, I just checked on it to see if I could reproduce what others saw.

If you're getting the crash still I'd recommend double checking that you've pull down the latest version of the branch (it's been rebased so you'll need to do a reset). I'm no longer able to reproduce this issue after my recent changes, so further investigation will require new information.

> I've also since noticed that several of the remote events no longer fire (on Android; iOS untested). Specifically I'm seeing RemotePlay, RemotePause and RemoteDuck never firing. Some messages above claim using RemotePlayPause works, but I haven't seen that one fire either.

I'll take a look at this.

dodgex commented:
> Using this in my app running Expo 53 and works great with the new architecture, great work!
> 
> I have one question about this commit though: [b11736b](https://github.com/doublesymmetry/react-native-track-player/commit/b11736b7f249698730845339c086b7dc722ca1ee)
> 
> It removes both the code emitting `MetadataTimedReceived` events as well as the deprecated `EventType.PlaybackMetadataReceived`. Because of this I'm having issues getting the in stream metadata, by adding back the timed metadata event listener on init and a simpler version of `handleAudioPlayerTimedMetadataReceived` removing the usage of the deprecated `EventType.PlaybackMetadataReceived` fixes the issue. See example below:
> 
> ```
> // Trackplayer init()
> player.event.receiveTimedMetadata.addListener(self, handleAudioPlayerTimedMetadataReceived)
> 
> func handleAudioPlayerTimedMetadataReceived(metadata: [AVTimedMetadataGroup]) {
>         let metadataItems = MetadataAdapter.convertToGroupedMetadata(metadataGroups: metadata);
>         emit(event: EventType.MetadataTimedReceived, body: ["metadata": metadataItems])
> }
> ```
> 
> I believe this might have been an unintended change, but not sure :)

@jspizziri I can confirm this issue and that adding back the code fixes it. would be awesome, if you could fix the timed metadata on iOS by partially reverting [b11736b](https://github.com/doublesymmetry/react-native-track-player/commit/b11736b7f249698730845339c086b7dc722ca1ee).

Thanks a lot. :)

doughsay commented:
> While trying to debug another notification related issue (on iOS), i tried the example app from the same revision/commit and the notification events dont work. My App is running on react-native `0.79.5` (without expo) cant upgrade to `0.80.x` yet. Maybe the issue is related to RN 0.80?

@dodgex I'm not sure I'm totally following. Are you claiming the events (RemotePlay, RemotePause, etc) _do_ fire in your 0.79.5 app but they're _not_ firing in the example app which uses 0.80.0?  There might be more to it then just the RN version, because I'm on 0.79.5 (expo 53) and they're not firing for me...

dodgex commented:
@doughsay Yes, this is exactly what I was claiming.

(Tested only for iOS; iPad Air 4th Gen with 18.5)

- My App. RN 0.79.5 with 4.0.1 as well as this banch -> working
- My App. RN 0.80.2 with this banch -> working
- Example from this branch as is -> not working *(see below)
- Example from this branch, but RN updated to 0.80.2 -> working

Now to the fun part, i wanted to reproduce the "not working" part for this branch, git clean'ed (-fdX) the example, reverted all my changes, uninstalled the example from my ipad, did a clean build -> it works... i have no fricking idea why.


But the issue I actually tried to debug when I tested the example is still present. For my App the Play button on the control center is grayed out when I start playback for the first time when my app starts, no idea why. Had it reproduced on the example for like 2-3 restarts but then it was "fixed" again... no success on my app tho.

lovegaoshi commented:
the bug reports are getting way too complicated for this PR and I recommend each of you open a **separate bug tracker**.

for example the timedMetadataReceived is NOT supposed to fire at all bc [it is removed](https://github.com/doublesymmetry/react-native-track-player/commit/b11736b7f249698730845339c086b7dc722ca1ee#diff-c4b3781bbc5e65b7067116b5213c1a607ed9ffa82e1b2d345c2743d4338d6078L737). if you do see it fire you are NOT using this branch.

lutfullo777 commented:
> @tux2nicolae the thing that's currently preventing me from merging and releasing this is the issue where `Remote` events aren't being fired on Android. Something seems to be wrong with the `MediaLibrarySession.Callback`'s `onMediaButtonEvent`, as it's simply never firing. I've spent some time investigating but haven't found a solution yet. If we could get that issue sorted out I think I'm ready to merge and release an alpha.
> 
> @lovegaoshi not sure if you have any insights on what might be causing this.

This doesn't work on iOS either. I checked with react native: 0.79.5. Everything else worked.

jspizziri commented:
@ovbm , I spun up a test project on react-native `0.79.5` and events are _NOT_ firing there for me either. Can you push a repo that uses this branch that does have events firing?

ovbm commented:
> @ovbm that repo must be private, as your link gives me a 404.

sorry, my bad. it's now public.

jspizziri commented:
> Turns out RemotePlay and RemotePause are working for me, but only if using a connected bluetooth headset 

@doughsay fascinating... I just confirmed this as well. This is potentially helpful information, thanks!

jspizziri committed:
fix(android): fix an issue causing remote events to not fire

jspizziri committed:
chore(example): upgrade example project to react-native 0.80.2

jspizziri committed:
chore(android): strip out android-auto specific code

jspizziri commented:
@revenkroz , this issue doesn't reproduce in the example app. There's something wrong on your end.

dodgex commented:
> Alright, I've fixed the issue where the [remote events weren't firing](https://github.com/doublesymmetry/react-native-track-player/pull/2478/commits/00018950f64f6a1bec814351d9bac20eccf22b28).
> 
> While I was looking around I found a lot of Android Auto related code that I'm going to strip back out of this PR for the next release. To reiterate, this release is going to focus specifically on media3 and turbomodules.
> 
> EDIT: I'm going to wait a few days for some community feedback on the remote events. If everything checks out, I plan on releasing a `5.0.0-alpha.0` soon.

Hey @jspizziri cool that you fixed the events! :)

Could you consinder https://github.com/doublesymmetry/react-native-track-player/pull/2478#issuecomment-3121464982 to fix `MetadataTimedReceived` events on iOS, before going `5.0.0-alpha.0`?

Thanks in advance! ❤️

doughsay commented:
I can confirm the remote events are now fixed for me!  Thanks! ❤️

jspizziri committed:
fix(ios): fix errors that occur when subscribing to the remote-play-pause event

jspizziri committed:
fix(ios): add back errantly removed MetadataTimedReceived event

h1rdr3v2 commented:
Please does this PR solve this issue

`WARN  The objective-c getSleepTimerProgress:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject method signature for the JS method getSleepTimerProgress can not be found in the ObjecitveC definition of the TrackPlayerModule module.
The getSleepTimerProgress JS method will not be available.`
` WARN  The objective-c setSleepTimer:(double)time resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject method signature for the JS method setSleepTimer can not be found in the ObjecitveC definition of the TrackPlayerModule module.
The setSleepTimer JS method will not be available.`
 `WARN  The objective-c sleepWhenActiveTrackReachesEnd:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject method signature for the JS method sleepWhenActiveTrackReachesEnd can not be found in the ObjecitveC definition of the TrackPlayerModule module.
The sleepWhenActiveTrackReachesEnd JS method will not be available.`
 `WARN  The objective-c clearSleepTimer:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject method signature for the JS method clearSleepTimer can not be found in the ObjecitveC definition of the TrackPlayerModule module.
The clearSleepTimer JS method will not be available.`

tomislav-arambasic committed:
fix(android): prevent livestream failures by assigning a default user-agent for Media3 requests (#2496)

Added a custom user-agent string to prevent 403 errors when playing Live365 streams using Media3.

leonardopicinin commented:
I’m testing the library on the latest commit (492d349) and here are some considerations that might help...

My setup:
React Native – 0.76.9
Expo – 52 managed

Android 16 and 15: Everything seems to work fine in the emulator, but on some real devices it doesn’t behave as expected, for example, when trying to use the remote controllers with the expanded notification (OnePlus 11 device). In other cases, the controller only closes if the app itself is closed.

Android 12 and 9: NOTHING works, none of the remote events.

iOS 18.5: Everything seems to be working very well.

tomislav-arambasic commented:
> @tomislav-arambasic 
> 
> 
> 
> > Could someone explain where we at with CarPlay & Android Auto?
> 
> 
> 
> While those issues are very important and a priority for many (including me), this PR specifically is leaving carplay and android auto alone. Right now we're making some needed upgrades to the library that will, ideally pave the way for those things. We can talk about the future of android auto and carplay elsewhere once this has been released and is stable.

Understandable. Appreciate the answer.

jspizziri committed:
build(package.json): update build commands

jspizziri committed:
v5.0.0-alpha0

jspizziri committed:
chore: resolve conflicts

mcalc001 commented:
Hi @jspizziri,

We’re still encountering a critical issue on Android with react-native-track-player.

Error:

```
Fatal Exception: java.lang.RuntimeException  
Unable to start service com.doublesymmetry.trackplayer.service.MusicService@ef92d81 with null:  
android.app.ForegroundServiceStartNotAllowedException: Service.startForeground() not allowed due to mAllowStartForeground false: service tv.liverpoolfc/com.doublesymmetry.trackplayer.service.MusicService
```

Environment:
	•	expo: ~51.0.39
	•	react-native: 0.74.5
	•	react-native-track-player: 4.1.1

Is this a known issue, or are there any recommended workarounds or fixes? This is currently our highest crash source on Android.

Thanks for your help 🙏

watadarkstar commented:
> [v5.0.0-alpha0](https://github.com/doublesymmetry/react-native-track-player/releases/tag/v5.0.0-alpha0) has been [published to npm](https://www.npmjs.com/package/react-native-track-player/v/5.0.0-alpha0).

@jspizziri Any idea when this will be stable? It still has bugs on Android like @mcalc001 pointed out.

cbdeveloper commented:
Should this be updated? It still says that it does not support the newArch.

https://reactnative.directory/package/react-native-track-player

gavrichards commented:
> Should this be updated? It still says that it does not support the newArch.
> 
> https://reactnative.directory/package/react-native-track-player

It doesn't. The last official release is still 4.1.2 which doesn't support the New Architecture. The only version of this library that does (although there are known unaddressed bugs) is an alpha version. Alpha versions are generally reserved for internal testing.

subvertallchris commented:
Is there a timeline for more updates to this and getting v5 stable and launched? We'd be happy to discuss funding some of this work if it helps.

RobSchilderr commented:
> Is there a timeline for more updates to this and getting v5 stable and launched? We'd be happy to discuss funding some of this work if it helps.

Also interested in this timeline / discuss solutions.
